src/network/networkd-dhcp4.c

```diff
@@ -571,6 +575,143 @@ static int dhcp_remove_dns_routes(Link *link, sd_dhcp_lease *lease, const struct
+        int r;
+
+        assert(link);
+        assert(link->dhcp4_messages > 0);
+
+        link->dhcp4_messages--;
+
+        if (IN_SET(link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+                return 1;
+
+        r = sd_netlink_message_get_errno(m);
+        if (r < 0 && r != -EEXIST) {
+                log_link_message_warning_errno(link, m, r, "Could not set DHCPv4 routing policy rule: %m");
+                link_enter_failed(link);
+                return 1;
+        }
+
+        if (link->dhcp4_messages == 0 && !link->network->dhcp_send_decline)
+                dhcp4_check_ready(link);
+        return 1;
+}
+
+static int dhcp4_add_source_routing_policy_rule(Link *link) {
+        struct in_addr address;
+        _cleanup_(routing_policy_rule_freep) RoutingPolicyRule *rule = NULL;
+        _cleanup_free_ char *from_addr_str = NULL;
+        int r;
+
+        assert(link);
+
+        if (!link->dhcp_lease || !link->network) /* link went down while we configured the IP addresses? */
+                return 0;
+
+        if (!link->network->dhcp_source_routing_enabled)
+                return 0;
+
+        r = sd_dhcp_lease_get_address(link->dhcp_lease, &address);
+        if (r < 0)
+                return log_link_warning_errno(link, r, "DHCP error: no address: %m");
+
+        r = routing_policy_rule_new(&rule);
+        if (r < 0)
+                return log_oom();
+
+        rule->from.in = address;
+        rule->from_prefixlen = 32;
+        rule->family = AF_INET;
+        rule->priority = link->network->dhcp_source_routing_rule_priority;
+        if (link->network->dhcp_route_table_set)
+                rule->table = link->network->dhcp_route_table;
+
+        if (DEBUG_LOGGING) {
+                const union in_addr_union address_union = { .in = address };
+                (void) in_addr_to_string(AF_INET, &address_union, &from_addr_str);
+        }
+
+        /* Only actually add the rule if we didn't already add the same rule from a
+        previous lease. Otherwise, routing_policy_rule_configure will error with -EEXIST */
+        if (
+                link->dhcp4_source_routing_policy_rule &&
+                routing_policy_rule_compare_func(link->dhcp4_source_routing_policy_rule, rule)
+        ) {
+                log_link_debug(link, "New and old source routing rules are the same, won't reconfigure");
+                return 0;
+        }
+        /* Now, if we really are going to set up a new routing policy rule, free the old one.
+        The dhcp4_remove_source_routing_policy_rule function will take care of actually calling
+        routing_policy_rule_remove based on the old lease address later on */
+        if (link->dhcp4_source_routing_policy_rule)
+                routing_policy_rule_free(TAKE_PTR(link->dhcp4_source_routing_policy_rule));
+
+
+        log_link_debug(
+                link, "Adding source routing rule %s/%d -> * table %d pri %d family %d",
```

Comment:
Yup, I can tidy this up - I'll also replace `*` with `0.0.0.0/0` since I think that's what's used by the debug logging in `routing_policy_rule_configure.